packages/flame_network_image/lib/src/flame_network_image.dart

```diff
@@ -0,0 +1,189 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:ui';
+
+import 'package:flame/cache.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:http/http.dart' as http;
+import 'package:path/path.dart' as path;
+import 'package:path_provider/path_provider.dart';
+
+/// {@template flame_image_response}
+/// A class containing the relevant http attributes to
+/// Flame Images Network package.
+/// {@endtemplate}
+class FlameImageResponse {
+  /// {@macro flame_image_response}
+  const FlameImageResponse({
+    required this.statusCode,
+    required this.bytes,
+  });
+
+  /// Http status code.
+  final int statusCode;
+
+  /// response bytes.
+  final Uint8List bytes;
+}
+
+/// Function signature used by Flame Network Images to fetch images.
+typedef GetImageFunction = Future<FlameImageResponse> Function(
+  String url, {
+  Map<String, String>? headers,
+});
+
+/// Function signature by Flame Network Images to get the app directory
+/// which is used for the local storage caching.
+typedef GetAppDirectoryFunction = Future<Directory> Function();
+
+/// {@template flame_network_images}
+///
+/// [FlameNetworkImages] is a class similar to Flame's [Images], but instead of
+/// loading images from the assets bundle, it loads from networks urls.
+///
+/// By default, [FlameNetworkImages] uses [http.get] method to make the requests
+/// ,that can be customized by passing a different [GetImageFunction] to the get
+/// argument on the constructor.
+///
+/// [FlameNetworkImages] also will also automatically cache files in a two layer
+/// system.
+///
+/// The first layer is an in memory cache, handled by an internal [MemoryCache],
+/// while the second is the device own file system, where images are cached in
+/// the application document directory, which by default is provided by
+/// path_providers' [getApplicationDocumentsDirectory] method, and can be
+/// customized using the getAppDirectory argument in the constructor.
+///
+/// When an image is requested, [FlameImageResponse] will first check on its
+/// cache layers before making the http request, if both layer are cache miss,
+/// then the request is made and both layers set with the response.
+///
+/// Another important note about the cache layers is that the first layer, is a
+/// per instance cache, while the local storage is a app global cache. This
+/// means that two different [FlameImageResponse] instances will have the same
+/// local storage cache, but not the memory cache.
+///
+/// Note that the local storage layer is not present when running on web since
+/// that platform doesn't really have a file system. The Browse caching will
+/// work as similar replacement for this layer, though that can't be controlled
+/// this package, make sure that the server where the images are being fetched
+/// returns the correct cache header to make the browser cache the assets.
```

Comment:
All modern browsers have `localStorage`, aka Web Storage (https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage), which can be used to persist files indefinitely, or at least until the user manually purges the cache.

Reply:
Yeah, I considered it, but I don't believe it is a good fit for what we are building here.

One of the biggest reason for that is that [`setItem`](https://developer.mozilla.org/en-US/docs/Web/API/Storage/setItem) values are strings, so we will need to serialize the assets into a base64 for example, something that can increase the size of the file when stored, and probably add some small CPU overhead.

In addition to that, local storage has a very limited storage capacity, I haven't any reference in the web with the correct value, but I remember from my old days as a JS developer that it was around 5mb.

So by relying on the browser cache, we have more storage capacity, code simplicity (since we don't need to write any JS interop code).

Reply:
Ah, you're right, it's only 5Mb. Though there's also [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) which has no size limit, but lasts only while the corresponding browser tab is open.

As for the string format - you're right again! So, some kind of encoding will be necessary. The base64 format uses 6 bits per byte, which means its overhead is 33%. However, the utf16 encoding (used by localStorage) has 63488 valid code points per 2-bytes, which means that theoretically there could be an encoding of bytes to utf16 with overhead of only 0.28%! But is there such a standard? Would we need to invent one?

However, all this is probably unnecessary for the initial version of the package, and perhaps never.